src/flagpole/evaluation_context.py

```diff
@@ -13,40 +13,78 @@
 EvaluationContextDict = dict[str, ValidContextTypes]
 
 
+class InvalidIdentityFieldException(Exception):
+    pass
+
+
 class EvaluationContext:
     """
     Prepared by the application and passed to flagpole to evaluate
     feature conditions.
     """
 
-    def __init__(self, data: EvaluationContextDict):
+    __data: EvaluationContextDict
+    __identity_fields: set[str]
+    id: int
+
+    def __init__(self, data: EvaluationContextDict, identity_fields: set[str] | None = None):
         self.__data = deepcopy(data)
 
-    def get(self, key: str) -> Any:
-        return self.__data.get(key)
+        if identity_fields is None:
+            # If no list of identity fields is provided, use all properties of
+            # the context.
+            self.__identity_fields = set(self.__data.keys())
+        else:
+            self.__identity_fields = deepcopy(identity_fields)
+            self.__validate_identity_fields()
 
-    def has(self, key: str) -> Any:
-        return key in self.__data
+        self.__id = self.__generate_id()
 
-    def size(self) -> int:
-        return len(self.__data)
+    def __validate_identity_fields(self):
+        invalid_identity_fields = []
 
-    def id(self) -> int:
+        for field in self.__identity_fields:
+            if field not in self.__data:
+                invalid_identity_fields.append(field)
+
+        if invalid_identity_fields:
+            raise InvalidIdentityFieldException(
+                f"One or more invalid identity fields specified: {invalid_identity_fields}"
+            )
+
```

Comment:
This logic probably has to go. Our evaluation context won't always contain project IDs, so I think we just have to allow it to skip over undefined fields and remove them from the identity fields set. Of course, this also means an empty ID set in any way should just use all keys as the ID fields, but this feels like really obscure behavior.

Reply:
Removing unknown fields from the identity_fields makes sense to me. If we end up with zero identity fields, then the all fields are part of the identity.

Reply:
Should be addressed now.